File: frontend/types/speed-curves.ts
export type SpeedCurvePreset = "hero" | "flash_in" | "montage" | "custom";
export interface SpeedCurvePoint { position: number; speed: number; }
export interface ClipSpeedCurve { clip_id: string; preset: SpeedCurvePreset; points: SpeedCurvePoint[]; }

export const SPEED_CURVE_PRESETS: Record<Exclude<SpeedCurvePreset, "custom">, Omit<ClipSpeedCurve, "clip_id">> = {
  hero: { preset: "hero", points: [{ position: 0, speed: 1 }, { position: .22, speed: 2.4 }, { position: .56, speed: .25 }, { position: .80, speed: 1.65 }, { position: 1, speed: 1 }] },
  flash_in: { preset: "flash_in", points: [{ position: 0, speed: 4.5 }, { position: .18, speed: 2.2 }, { position: .48, speed: 1 }, { position: 1, speed: 1 }] },
  montage: { preset: "montage", points: [{ position: 0, speed: 1 }, { position: .22, speed: 2.7 }, { position: .45, speed: .7 }, { position: .70, speed: 2.2 }, { position: 1, speed: 1.15 }] },
};

export function speedToGraphY(speed: number): number {
  return 1 - (Math.log10(Math.max(.1, Math.min(10, speed))) + 1) / 2;
}

export function graphYToSpeed(y: number): number {
  return Math.max(.1, Math.min(10, 10 ** (1 - 2 * Math.max(0, Math.min(1, y)))));
}

File: backend/app/schemas/speed_curves.py
from __future__ import annotations

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class SpeedCurvePoint(BaseModel):
    position: float = Field(ge=0, le=1, description="Normalised position within the source clip")
    speed: float = Field(ge=.1, le=10)


class ClipSpeedCurve(BaseModel):
    clip_id: UUID
    preset: Literal["hero", "flash_in", "montage", "custom"] = "custom"
    points: list[SpeedCurvePoint] = Field(min_length=2, max_length=12)

    @model_validator(mode="after")
    def complete_curve(self) -> "ClipSpeedCurve":
        if self.points[0].position != 0 or self.points[-1].position != 1:
            raise ValueError("Speed curve must begin at 0 and end at 1")
        if any(right.position <= left.position for left, right in zip(self.points, self.points[1:])):
            raise ValueError("Speed curve points must be strictly ordered")
        return self


class TimelineSpeedCurveUpdateRequest(BaseModel):
    curves: list[ClipSpeedCurve] = Field(min_length=1, max_length=100)


class TimelineSpeedCurveUpdateResponse(BaseModel):
    timeline_id: UUID
    status: str
    curves: list[dict]

